### PR TITLE
feat: PHPCS: adding support for the WordPress VIP Go standard

### LINKED ISSUE #1535: Introduce WordPress VIP coding standards and fixes for existing errors
WordPress VIP has pretty high coding standards and everyone who will work with VIP in the future will face the same problems as our team right now. It would be great if we could get coding standards up and running with VIP's standards.

Ideally we could setup PHP_Codesniffer with all the configurations of them and let a bot check run through it before merging PRs.

We will attempt to setup everything and also fix major errors, so that we will be able to use WPGraphQL in the future. It would be great if some maintainers can share some of their attention on this, so we don't end up breaking stuff.

Also this is obviously open for discussion, we would just use that PR for our purposes anyways, but would greatly appreciate if it would make it into the main repos best practices.

**References:** 
- https://wpvip.com/documentation/how-to-install-php-code-sniffer-for-wordpress-vip/
- https://github.com/Automattic/VIP-Coding-Standards/tree/master/WordPress-VIP-Go

### PR DESCRIPTION
**Pull request by renatonascalves**
What does this implement/fix? Explain your changes.
---------------------------------------------------
- Introduces support for the WordPress VIP Go standard
- It either silences or fixes the errors
- It removes duplicate composer packages already bundled in `automattic/vipwpcs`

Does this close any currently open issues?
------------------------------------------
fixes #1535 

Any relevant logs, error output, GraphiQL screenshots, etc?
-------------------------------------
(If it’s long, please paste to https://ghostbin.com/ and insert the link here.)

Any other comments?
-------------------
Code reviewer: see my comments for some questions/considerations.

### CODE REVIEW

File: src/Data/NodeResolver.php [hunk 1]
```diff
@@ -207,7 +207,7 @@ public function resolve_uri( string $uri, $extra_query_vars = '' ) {
 
 						if ( $wp_rewrite->use_verbose_page_rules && preg_match( '/pagename=\$matches\[([0-9]+)\]/', $query, $varmatch ) ) {
 							// This is a verbose page match, let's check to be sure about it.
-							$page = get_page_by_path( $matches[ $varmatch[1] ] );
+							$page = get_page_by_path( $matches[ $varmatch[1] ] ); // phpcs:ignore WordPressVIPMinimum.Functions.RestrictedFunctions.get_page_by_path_get_page_by_path
```
Comment (renatonascalves) [inline src/Data/NodeResolver.php:210]: For this one, should we do a if/else strategy with `wpcom_vip_get_page_by_path`?

File: src/Router.php [hunk 2]
```diff
@@ -392,17 +392,18 @@ public static function set_headers() {
...
-	 * @global string php://input Raw post data.
 	 * @since  0.0.5
+	 *
+	 * @global string php://input Raw post data.
+	 *
+	 * @return string|false Raw request data.
 	 */
 	public static function get_raw_data() {
-		$input = file_get_contents( 'php://input' );
+		$input = file_get_contents( 'php://input' ); // phpcs:ignore WordPressVIPMinimum.Performance.FetchingRemoteData.FileGetContentsRemoteFile
```
Comment (renatonascalves) [inline src/Router.php:402]: VIP recommends:

> `file_get_contents()` is highly discouraged for remote requests, please use `wpcom_vip_file_get_contents()` or `vip_safe_wp_remote_get()`

File: src/Type/ObjectType/RootQuery.php [hunk 3]
```diff
@@ -683,11 +683,11 @@ public static function register_post_object_fields() {
 								break;
 							case 'source_url':
 								$url     = $args['id'];
-								$post_id = attachment_url_to_postid( $url );
+								$post_id = attachment_url_to_postid( $url ); // phpcs:ignore WordPressVIPMinimum.Functions.RestrictedFunctions.attachment_url_to_postid_attachment_url_to_postid
```
Comment (renatonascalves) [inline src/Type/ObjectType/RootQuery.php:686]: Another one where we could do a if/else strategy with `wpcom_vip_attachment_url_to_postid`

**Review by justlevine — COMMENTED** [review]
Awesome work @renatonascalves !

Re adopting `vipcom_` functions, not saying we should, but _if_ we do: 
- let's wrap the `if( method_exists( 'vipcom_fn') {return vipcom_fn();} return wp_fn();` logic in a separate class (`Utils\Plugabble` or something) to keep things DRY.
*Edit:* we also might want to add a hook, or drop the namespace and wrap each method in its own `method_exists()`, so we can provide the same first-party parity with non-WPVip setups or plugins like `wp-graphql-smart-cache`
- A lot of those methods employ caching, so we should make sure they're unit tested to avoid issues we had with the ObjectCursor in #2457 
- Anything in `NodeResolver::resolve_uri()` should be left untouched (ie errors suppressed, no code change) until #2342 is merged. Currently most of it is lifted directly from `WP:parse_request()`, isn't unit tested, and is a source for a plethora of bugs ( 👀 #2366 ), so we should avoid changing it unless absolutely necessary.

**Comment by renatonascalves** [comment]
@justlevine May I suggest then to move on silencing those examples in this issue? I'll create issues to address those separate changes.

Introducing unit tests, for example, would increase the scope a little bit since we would need to add support for the VIP `mu-plugins` into the unit tests and write the tests themselves.

**Comment by renatonascalves** [comment]
@jasonbahl Could you code review this pull request when you have a chance?

**Review by jasonbahl — APPROVED** [review]

**Comment by jasonbahl** [comment]
@renatonascalves I think I agree with @justlevine about adding vip support in a separate file. 

If any of these functions can't be filtered and replaced with the VIP equivalent, it's probably a good candidate for a new filter. 

Ideally, if do need to introduce new filters, I'd like to keep them central when possible. 

For example (doesn't apply here, but just to set the tone), if we needed to filter something in the UserLoader, if possible I'd prefer the `apply_filters()` call to exist in the AbstractDataLoader, not the UserLoader. That way _any_ loader could use the filter, not _just_ the user loader. (sometimes it can't be centralized, and that's ok, but when it can be, that's my preference).

**Comment by renatonascalves** [comment]
Thank you! I'll proceed to merge this and create issues for the follow-up work.

Do you guys know what's this **Check code (8.0.15)** Github Action about?

**Comment by jasonbahl** [comment]
@renatonascalves I think  I need to update the required checks. I think it's tied to an old PHP version that was updated. Looking now 👀

**Comment by jasonbahl** [comment]
@renatonascalves ya, it was related to a change on the Github Workflow that checks PHPStan. 

I updated the branch settings. Good to go now.

**Comment by renatonascalves** [comment]
Thank you, Jason!

**Pull request merged**